firedrake/interpolation.py

```diff
@@ -289,30 +273,24 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
     parameters = {}
     parameters['scalar_type'] = utils.ScalarType
 
-    if not isinstance(expr, firedrake.Expression):
-        kernel = compile_expression_dual_evaluation(expr, to_element,
-                                                    domain=source_mesh,
-                                                    parameters=parameters)
-        ast = kernel.ast
-        oriented = kernel.oriented
-        needs_cell_sizes = kernel.needs_cell_sizes
-        coefficients = kernel.coefficients
-        first_coeff_fake_coords = kernel.first_coefficient_fake_coords
-        name = kernel.name
-        kernel = op2.Kernel(ast, name, requires_zeroed_output_arguments=True,
-                            flop_count=kernel.flop_count)
-    elif hasattr(expr, "eval"):
-        to_pts = []
-        for dual in to_element.fiat_equivalent.dual_basis():
-            if not isinstance(dual, FIAT.functional.PointEvaluation):
-                raise NotImplementedError("Can only interpolate Python kernels with Lagrange elements")
-            pts, = dual.pt_dict.keys()
-            to_pts.append(pts)
-        kernel, oriented, needs_cell_sizes, coefficients = compile_python_kernel(expr, to_pts, to_element, V, source_mesh.coordinates)
-        first_coeff_fake_coords = False
-    else:
-        raise RuntimeError("Attempting to evaluate an Expression which has no value.")
-
+    # We need to pass both the ufl element and the finat element
+    # because the finat elements might not have the right mapping
+    # (e.g. L2 Piola, or tensor element with symmetries)
+    # FIXME: for the runtime unknown point set (for cross-mesh
+    # interpolation) we have to pass the finat element we construct
+    # here. Ideally we would only pass the UFL element through.
```

Comment:
I think all the `to_element` creation (and any fiddling with `rebuild`) in `firedrake/interpolation.py` and `firedrake/preconditioners/pmg.py` can be absorbed into `compile_expression_dual_evaluation` as long as we work out how to tell `compile_expression_dual_evaluation` that it needs to make a runtime unknown point set for cross-mesh interpolation.